torchao/quantization/qat/api.py

```diff
@@ -11,19 +11,213 @@
+        )
+        from torchao.quantization.qat import QATConfig
+
+        base_config = Int8DynamicActivationInt4WeightConfig(group_size=32)
+        quantize_(model, QATConfig(base_config, step="prepare"))
+        train_loop(model)
+        quantize_(model, QATConfig(base_config, step="convert"))
+
+    Currently only the following are supported as base configs:
+
+        - :class:`~torchao.quantization.Int8DynamicActivationInt4WeightConfig`
+        - :class:`~torchao.quantization.Int4WeightOnlyConfig`
+
+    The second way to use this config involves specifying the fake quantization
+    schemes directly. Users will pass in :class:`~torchao.quantization.qat.FakeQuantizeConfigBase`
+    for weights and/or activations instead of the base PTQ config. This use case
+    is mostly for experimentation, e.g. when the corresponding PTQ config does
+    not exist yet.
+
+    Example usage::
+
+        from torchao.quantization import quantize_
+        from torchao.quantization.qat import IntxFakeQuantizeConfig
+
+        activation_config = IntxFakeQuantizeConfig(
+            torch.int8, "per_token", is_symmetric=False,
+        )
+        weight_config = IntxFakeQuantizeConfig(
+            torch.int4, group_size=32, is_symmetric=True,
+        )
+        qat_config = QATConfig(
+            # must specify one of `base_config` or `weight_config`
+            activation_config=act_config,
+            weight_config=weight_config,
+            step="prepare",
+        )
+        quantize_(model, qat_config)
+
+    Args:
+        base_config (Optional[AOBaseConfig]): Base PTQ config to infer the fake
+            quantization configs during the prepare phase, and to apply directly
+            during the convert phase.
+        activation_config (Optional[FakeQuantizeConfigBase]): Custom fake
+            quantization config for input activations, always optional.
+            Must be None if `base_config` is used.
+        weight_config (Optional[FakeQuantizeConfigBase]): Custom fake quantization
+            config for weights. Must be None if `base_config` is used.
+
+    Keyword args:
+        step (str): One of "prepare" or "convert", determines the QAT phase
+
+    Raises:
+        ValueError: If `base_config` and `activation_config` are both specified
+        ValueError: If `base_config` and `weight_config` are both specified
+        ValueError: If neither `base_config` nor `weight_config` is specified
+        ValueError: If `step` is not one of "prepare" or "convert"
+        ValueError: If `base_config` is None but `step` is "convert"
+        ValueError: If the config is applied on a module that is not a
+            `torch.nn.Linear` or `torch.nn.Embedding`, or it is applied on
+            `torch.nn.Embedding` with an activation config
+    """
+
+    base_config: Optional[AOBaseConfig]
+    activation_config: Optional[FakeQuantizeConfigBase]
+    weight_config: Optional[FakeQuantizeConfigBase]
+    step: str
```

Comment:
nit: make a string enum, and accept the user argument as both a string (no change) and an enum value